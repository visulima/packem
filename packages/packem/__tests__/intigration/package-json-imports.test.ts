import { rm } from "node:fs/promises";

import { readFileSync, writeFileSync } from "@visulima/fs";
import { temporaryDirectory } from "tempy";
import { afterEach, beforeEach, describe, expect, it } from "vitest";

import { createPackageJson, createPackemConfig, createTsConfig, execPackemSync, installPackage } from "../helpers";

describe("packem package.json imports", () => {
    let temporaryDirectoryPath: string;

    beforeEach(async () => {
        temporaryDirectoryPath = temporaryDirectory();
    });

    afterEach(async () => {
        await rm(temporaryDirectoryPath, { recursive: true });
    });

    it.each(["esbuild", "swc", "sucrase"] as ("esbuild" | "swc" | "sucrase")[])(
        "should resolve package.json imports with a * and %s transformer",
        async (transformer) => {
            expect.assertions(5);

            writeFileSync(`${temporaryDirectoryPath}/src/x.ts`, `export const x = 2;`);
            writeFileSync(`${temporaryDirectoryPath}/src/index.ts`, `export * as warns from '#x.ts';`);

            await installPackage(temporaryDirectoryPath, "typescript");

            await createPackageJson(
                temporaryDirectoryPath,
                {
                    devDependencies: {
                        typescript: "*",
                    },
                    exports: {
                        ".": {
                            import: "./dist/index.mjs",
                            require: "./dist/index.cjs",
                        },
                    },
                    imports: {
                        "#*": "./src/*",
                    },
                },
                transformer,
            );
            await createTsConfig(temporaryDirectoryPath);
            await createPackemConfig(temporaryDirectoryPath, {
                transformer,
            });

            const binProcess = await execPackemSync("build", [], {
                cwd: temporaryDirectoryPath,
            });

            expect(binProcess.stderr).toBe("");
            expect(binProcess.exitCode).toBe(0);

            expect(binProcess.stdout).not.toContain("If this is incorrect, add it to the");

            const cjsContent = readFileSync(`${temporaryDirectoryPath}/dist/index.cjs`);

            expect(cjsContent).toMatchSnapshot("cjs output");

            const mjsContent = readFileSync(`${temporaryDirectoryPath}/dist/index.mjs`);

            expect(mjsContent).toMatchSnapshot("mjs output");
        },
    );
});
