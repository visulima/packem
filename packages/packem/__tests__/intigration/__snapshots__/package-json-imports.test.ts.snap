// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`packem package.json imports > should resolve package.json imports with a * and esbuild transformer > cjs output 1`] = `
"'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const x = 2;

const x$1 = /*#__PURE__*/Object.defineProperty({
	__proto__: null,
	x
}, Symbol.toStringTag, { value: 'Module' });

exports.warns = x$1;
"
`;

exports[`packem package.json imports > should resolve package.json imports with a * and esbuild transformer > mjs output 1`] = `
"const x = 2;

const x$1 = /*#__PURE__*/Object.defineProperty({
	__proto__: null,
	x
}, Symbol.toStringTag, { value: 'Module' });

export { x$1 as warns };
"
`;

exports[`packem package.json imports > should resolve package.json imports with a * and oxc transformer > cjs output 1`] = `
"'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const x = 2;

const x$1 = /*#__PURE__*/Object.defineProperty({
	__proto__: null,
	x
}, Symbol.toStringTag, { value: 'Module' });

exports.warns = x$1;
"
`;

exports[`packem package.json imports > should resolve package.json imports with a * and oxc transformer > mjs output 1`] = `
"const x = 2;

const x$1 = /*#__PURE__*/Object.defineProperty({
	__proto__: null,
	x
}, Symbol.toStringTag, { value: 'Module' });

export { x$1 as warns };
"
`;

exports[`packem package.json imports > should resolve package.json imports with a * and sucrase transformer > cjs output 1`] = `
"'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

const x = 2;

const x$1 = /*#__PURE__*/Object.defineProperty({
	__proto__: null,
	x
}, Symbol.toStringTag, { value: 'Module' });

exports.warns = x$1;
"
`;

exports[`packem package.json imports > should resolve package.json imports with a * and sucrase transformer > mjs output 1`] = `
"const x = 2;

const x$1 = /*#__PURE__*/Object.defineProperty({
	__proto__: null,
	x
}, Symbol.toStringTag, { value: 'Module' });

export { x$1 as warns };
"
`;

exports[`packem package.json imports > should resolve package.json imports with a * and swc transformer > cjs output 1`] = `
"'use strict';

Object.defineProperty(exports, Symbol.toStringTag, { value: 'Module' });

var x = 2;

const x$1 = /*#__PURE__*/Object.defineProperty({
	__proto__: null,
	x
}, Symbol.toStringTag, { value: 'Module' });

exports.warns = x$1;
"
`;

exports[`packem package.json imports > should resolve package.json imports with a * and swc transformer > mjs output 1`] = `
"var x = 2;

const x$1 = /*#__PURE__*/Object.defineProperty({
	__proto__: null,
	x
}, Symbol.toStringTag, { value: 'Module' });

export { x$1 as warns };
"
`;
